DirectProgramming/DPC++FPGA/include/streaming_cholesky_inversion.hpp

```diff
@@ -0,0 +1,263 @@
+
+namespace fpga_linalg {
+
+/*
+  Cholesky decomposition - Computes L such that A=LL* where:
+  - A is the input matrix (hermitian, positive definite)
+  - L is a lower triangular matrix
+  - L* is the conjugate transpose of L
+
+  This function implements a modified version of the Cholesky–Banachiewicz
+  algorithm.
+  Pseudo code:
+
+  int row_size = 0;
+  for (column = 0; column <= row_size; column++) {
+    for (row = column; row < rows; row++) {
+      float sum = 0;
+      for (k = 0; k < column; k++)
+        sum += L[row][k] * L[column][k];
+
+      if (row == column)
+        L[row][column] = sqrt(A[row][row] - sum);
+      else
+        L[row][column] = (A[row][column] - sum) / L[column][column];
+    }
+  }
+
+  The input and output matrices are consumed/produced from/to pipes.
+*/
+template <typename T,       // The datatype for the computation
+          bool is_complex,  // True if T is ac_complex<X>
+          int rows,         // Number of rows==columns in the A matrices
+          int raw_latency,  // Read after write (RAW) latency (in iterations) of
+                            // the triangular loop of this function.
+                            // This value depends on the FPGA target, the
+                            // datatype, the target frequency, etc.
+                            // This value will have to be tuned for optimal
+                            // performance. Refer to the Triangular Loop
+                            // design pattern tutorial.
+                            // In general, find a high value for which the
+                            // compiler is able to achieve an II of 1 and
+                            // go down from there.
+          int pipe_size,    // Number of elements read/write per pipe operation
+                            // to read the input matrix
+          typename LIn,     // A matrix input pipe, receive pipe_size
+                            // elements from the pipe with each read
+          typename IOut     // L matrix output pipe, send one elements to the
+                            // pipe with each write.
+                            // Only lower-left elements of L are
+                            // sent in row order, starting with row 0.
+          >
+struct StreamingCholeskyInversion {
+  void operator()() const {
+    // Functional assertions
+    static_assert(rows >= 4,
+                  "Only matrices of size 4x4 and over are supported");
+    static_assert(pipe_size >= 1,
+                  "The pipe must be able to contain at least one element");
+
+    // Set the computation type to T or ac_complex<T> depending on the value
+    // of is_complex
+    using TT = std::conditional_t<is_complex, ac_complex<T>, T>;
+
+    constexpr int kColumns = rows;
+
+    // Compute Cholesky-based inversions as long as L input matrices are given
+    while (1) {
+      // L matrix read from pipe
+      [[intel::private_copies(4)]]  // NO-FORMAT: Attribute
```

Comment:
Maybe worth a comment here to explain why these private_copies and max_replicates values were chosen?